macaroonbakery/bakery.py

```diff
@@ -0,0 +1,280 @@
+        @param first_party_location the location of the first party.
+        @param primary_macaroon used for signature of the discharge macaroon.
+        @raise DischargeError when an error occurs during the discharge
+            process.
+        '''
+        caveats = macaroon.third_party_caveats()
+        for caveat in caveats:
+            dm = self._get_discharge(first_party_location,
+                                     caveat[0],
+                                     caveat[1])
+            dm = primary_macaroon.prepare_for_request(dm)
+            discharges.append(dm)
+            self._discharge_caveats(dm, discharges, first_party_location,
+                                    primary_macaroon)
+
+    def _get_discharge(self, location, third_party_location, condition):
+        ''' Get the discharge macaroon from the third party location.
+
+        @param location the origin.
+        @param third_party_location where to get a discharge from.
+        @param condition associated  to the discharged macaroon.
+        @return a discharged macaroon.
+        @raise DischargeError when an error occurs during the discharge
+            process.
+        '''
+        headers = {'Bakery-Protocol-Version': 1}
+        if self._discharge_token is not None:
+            headers['Macaroons'] = self._discharge_token
+        payload = {'id': condition, 'location': location}
+
+        try:
+            response = requests.post(third_party_location + '/discharge',
+                                     headers=headers,
+                                     data=payload,
+                                     timeout=self._timeout)
+        except requests.RequestException as exc:
+            raise DischargeError(
+                'Unable to access discharge endpoint: {}'.format(exc.message))
+        if response.status_code == 200:
+            return self._extract_macaroon_from_response(response)
+        elif response.status_code == 401:
+            visit_url, wait_url = self._extract_urls(response)
+            # Open a browser so the user can validate its identity.
+            webbrowser.open(visit_url, new=1)
+
+            # Wait on the wait url and then get a macaroon if validated.
+            return self._acquire_macaroon_from_wait(wait_url)
+        else:
+            raise DischargeError(
+                'Unknown status code from discharge endpoint: {}'.format(
+                    response.status_code)
+            )
+
+    def _extract_macaroon_from_response(self, response):
+        ''' Extarct the macaroon from a direct successful discharge.
+
+        @param response from direct successful discharge.
+        @return a macaroon object.
+        @raises DischargeError if any error happens.
+        '''
+        try:
+            response_json = response.json()
+        except ValueError:
+            raise DischargeError(
+                'Unable to access data from discharge endpoint on 200')
```

Comment:
Cannot decode JSON response from discharge endpoint ?

Also might be good to include the actual ValueError rather than throwing it away?